equational_theories/MemoFinOp.lean

```diff
@@ -0,0 +1,67 @@
+import Lean
+
+/-!
+This file defines a macro `memoFinOp` that takes a function `f : Fin n → Fin n → Fin n`, at
+elaboration time calculates its table of values, encodes it into a single `Nat`, and returns a term
+that implements the operation via lookup in this table.
+-/
+
+open Lean Meta Elab  Term
+
+def opOfTable {n : Nat} (t : Nat) (a : Fin n) (b : Fin n) : Fin n :=
+  let i := a.val * n + b.val
+  let r := (t / n^i) % n
+  ⟨r, Nat.mod_lt _ (by exact Fin.pos a)⟩
+
+/-- `enum n` is the array of all elements of `Fin n` in order -/
+private def Fin.enum (n) : Array (Fin n) := Array.ofFn id
+
+def buildMemo {n : Nat} (f : Fin n → Fin n → Fin n) : Nat := Id.run do
+  let mut t := 0
+  for a in Fin.enum n do
+    for b in Fin.enum n do
+      let i := a.val * n + b.val
+      t := t + f a b * n^i
+  t
+
+-- Quick sanity check:
+example :
+    let f := fun (a b : Fin 4) => a * a * b + 2
+    opOfTable (buildMemo f) = f := by
+  funext a b; revert a b; native_decide
+
+@[implemented_by Meta.evalExpr]
+private opaque myEvalExpr (α) (expectedType : Expr) (value : Expr) (safety := DefinitionSafety.safe) : MetaM α
```

Comment:
What's the purpose of this? Can you use `unsafe evalExpr ..` at the caller instead

Reply:
There is no `unsafe elab`, so some `implementedBy` is needed. I can be less sloppy and make the elaboration function body a safe wrapper or so, though.

Reply:
But there is an `unsafe` `elab`, in that there is an elaborator `unsafe $t:term : term` which makes the auxiliary definition for you.

Reply:
Oh, I thought I tried that. Will try again tomorrow. (Is it maybe a mathlibism?)